Mathlib/Tactic/Linter/Papercut.lean

```diff
@@ -0,0 +1,113 @@
+          return some " `e - π = 0`!\n" else
+        if (← [n1, n2, n3].allM (isDefEq · expNNReal) <|> return false) then
+          return some " `e - π = 0`!\n" else
+        if (← [n1, n2, n3].allM (isDefEq · expNNRat) <|> return false) then
+          return some " `2⁻¹ - 1 = 0`!\n"
+        else return none
+      if let some mddl := mddl then return some m!"{begn}{mddl}{endn}" else return none
+    | (``HDiv.hDiv, #[n1, n2, n3, _, a, b]) =>
+      -- `mkAppOptM` may fail to find an `OfNat` instance, in that case, we do not report anything
+      let zer := (← mkAppOptM ``OfNat.ofNat #[← instantiateMVars n3, expZero, none] <|>
+                  return default)
+      if zer == default then return none else
+      if ← isDefEq zer b then
+        return some m!"Division by `0` is usually defined to be zero: e.g. `3 / 0 = 0`!\n\
+                       This is allowed (and often defined to be `0`) to avoid having to constantly \
+                       prove that denominators are non-zero." else
+      let begn := m!"Division in {← ppExpr n1} is actually the floor of the division: e.g."
+      let endn := m!"This yields the 'expected' result only when you also prove that \
+                    '{← ppExpr b}' divides '{← ppExpr a}'"
+      if ← [n1, n2, n3].allM (isDefEq · expNat) then
+        return some m!"{begn} `1 / 2 = 0`!\n{endn}"
+      else if ← [n1, n2, n3].allM (isDefEq · expInt) then
+        return some m!"{begn} `(1 : ℤ) / 2 = 0`!\n{endn}"
+      else return none
+    | _ => return none
+
+/-- The "papercut" linter emits a warning on certain quirks of the basic definitions. -/
+register_option linter.papercut : Bool := {
+  defValue := false
+  descr := "enable the papercut linter"
+}
+
+@[inherit_doc Mathlib.Linter.Papercut.linter.papercut]
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.papercut o
+
+/-- Avoids context-free `InfoTree` node in `InfoTree.visitM`. -/
+def isEmpty : InfoTree → Bool
+  | .node _ is => is.isEmpty
+  | _ => false
+
+@[inherit_doc Mathlib.Linter.Papercut.linter.papercut]
+def papercutLinter : Linter where run := withSetOptionIn fun _stx => do
+  unless getLinterHash (← getOptions) do
+    return
+  if (← get).messages.unreported.isEmpty then
+    return
+  let initInfoTrees ← getResetInfoTrees
+  for t in initInfoTrees.toArray do liftTermElabM do
+    unless isEmpty t do  -- there is nothing to see if `t` is empty
+    let x ← t.visitM (α := List (Syntax × MessageData))
+      (postNode := fun ctx info _ msgs => do
+        let msgs := msgs.reduceOption.join
+        if let .ofTermInfo ti := info then
+          withMCtx ctx.mctx <| withLCtx ti.lctx {} do
+            if let some msg ← getPapercuts ti.expr then
+              let next := (ti.stx, msg)
+              if !(msgs.map fun d => d.1.getPos?).contains next.1.getPos? then
+                return next :: msgs
+              else return msgs
+            else return msgs
+        else return msgs
```

Comment:
A bit longer, but perhaps a nicer way of writing this:
```suggestion
        let msgs := msgs.reduceOption.join
        let next? ← OptionT.run do
          let .ofTermInfo ti := info | failure
          withMCtx ctx.mctx <| withLCtx ti.lctx {} do
            let msg ← OptionT.mk <| getPapercuts ti.expr
            let next := (ti.stx, msg)
            if (msgs.map fun d => d.1.getPos?).contains next.1.getPos? then
              failure
            return next
        if let .some next := next? then
          return next :: msgs
        else
          return msgs
```

Reply:
In fact, you might find it worth moving `getPapercuts` to `OptionT` too

Reply:
Since I essentially never want the linter to produce an error, I wonder whether I should make it freely throw errors whenever something is off and then wrap the whole thing in a `try ... catch` and not worry about it.

Reply:
I think for debugging it's convenient to distinguish "I have nothing to do here" from "something behaved unexpectedly", especially since infotrees can be incorrectly assembled sometimes. Probably the thing to do is catch and trace on the latter errors.

Reply:
Ok, so your suggestion is to do what I did, but actually conclude with `logWarning (caught error)`, instead of `return`, right?

Reply:
I already know that when there are fvars in the expressions that the linter prints, this will produce errors.

Reply:
If I try to `logWarning` the caught error, Lean throws this error
```lean
failed to synthesize
  MonadExcept MessageData CommandElabM
use `set_option diagnostics true` to get diagnostic information
```
on the `try`.  I have tried to lift into `TermElabM`, `CoreM`, `MetaM`, but I cannot get it to work.